File: Problem2.py
#Time Complexity :O(8m*n) 
#Space Complexity :O(m*n)
#Did this code successfully run on Leetcode : Yes
#Any problem you faced while coding this : No

from collections import deque
class Solution:
    def updateBoard(self, board: List[List[str]], click: List[int]) -> List[List[str]]:
        if not board:
            return board
        m=len(board)
        n=len(board[0])
        self.dirs=[(-1,-1),(0,-1),(1,-1),(-1,0),(1,0),(-1,1),(0,1),(1,1)]
        if board[click[0]][click[1]]=='M':
            board[click[0]][click[1]]='X'
            return board
        q=deque()
        q.append(click)
        board[click[0]][click[1]]='B'
        while q:
            curr=q.popleft()
            count=self.countMines(board,curr[0],curr[1])
            if count==0:
                for dir in self.dirs:
                    nr=dir[0]+curr[0]
                    nc=dir[1]+curr[1]
                    if nr>=0 and nc>=0 and nr<len(board) and nc<len(board[0]) and board[nr][nc]=='E':
                        q.append((nr,nc))
                        board[nr][nc]='B'
            else:
                board[curr[0]][curr[1]]=str(count)
        return board

    def countMines(self,board,i,j):
        count=0
        for d in self.dirs:
            nr=d[0]+i
            nc=d[1]+j
            if nr>=0 and nc>=0 and nr<len(board) and nc<len(board[0]) and board[nr][nc]=='M':
                count+=1
        return count




        

File: Problem_2.js
// Problem 2 Snakes and ladders (https://leetcode.com/problems/snakes-and-ladders/)

// Time Complexity : O(n^2)
// Space Complexity : O(n^2)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Flattening 2d array


// Your code here along with comments explaining your approach

/**
 * @param {number[][]} board
 * @return {number}
 */
var snakesAndLadders = function (board) {
    if (board === null || board.length === 0)
        return 0;
    let n = board.length;
    let nums = new Array(n * n);
    // Flatten the 2d array
    let row = n - 1, col = 0, even = 0;
    for (let i = 0; i < n * n; i++) {
        if (board[row][col] === -1) {
            nums[i] = -1;
        } else {
            nums[i] = board[row][col] - 1;
        }
        if (even % 2 === 0) {
            col++;
            if (col === n) {
                col--;
                row--;
                even++;
            }
        } else {
            col--;
            if (col === -1) {
                col++;
                even++;
                row--;
            }
        }
    }

    let queue = [];
    queue.push(0);
    nums[0] = -2;
    let level = 0;
    while (queue.length > 0) {
        let size = queue.length;
        for (let j = 0; j < size; j++) {
            curr = queue.shift();
            if (curr === ((n * n) - 1))
                return level;
            for (let i = 1; i <= 6; i++) {
                let newcurr = curr + i;
                if (newcurr >= n * n)
                    continue;
                if (nums[newcurr] === -1) {
                    queue.push(newcurr);
                    nums[newcurr] = -2;
                } else {
                    if (nums[newcurr] != -2) {
                        queue.push(nums[newcurr]);
                        nums[newcurr] = -2;
                    }
                }
            }
        }
        level++;
    }
    return -1;
};